File: plugins/vitest-agent-observability/src/judges/prompt.ts
/**
 * Build the combined LLM judge prompt for one or more rubrics.
 *
 * Sync target:
 *   plugins/pytest-agent-observability/src/pytest_agent_observability/judges/prompt.py
 */

import type { Rubric } from "./rubrics.js";
import type { JudgeInput } from "./runner.js";

export function buildPrompt(rubrics: Rubric[], input: JudgeInput): string {
  const sections: string[] = [];

  sections.push(
    "You are an objective evaluator. Assess the agent response below against each " +
      "rubric criterion and return ONLY a JSON object. No prose outside the JSON.\n" +
      "Response format:\n" +
      '{"<rubric_name>": {"score": <float 0.0-1.0>, "reason": "<brief explanation>"}, ...}',
  );

  // Input slots — omit empty
  if (input.systemPrompt) {
    sections.push(`<system_prompt>\n${input.systemPrompt}\n</system_prompt>`);
  }

  if (input.taskInstructions) {
    sections.push(
      `<task_instructions>\n${input.taskInstructions}\n</task_instructions>`,
    );
  }

  if (input.context) {
    const ctx = Array.isArray(input.context)
      ? input.context.join("\n")
      : input.context;
    sections.push(`<context>\n${ctx}\n</context>`);
  }

  if (input.conversationHistory && input.conversationHistory.length > 0) {
    const historyText = JSON.stringify(input.conversationHistory, null, 2);
    sections.push(
      `<conversation_history>\n${historyText}\n</conversation_history>`,
    );
  }

  sections.push(`<agent_response>\n${input.response}\n</agent_response>`);

  // Rubric definitions
  for (const rubric of rubrics) {
    const stepsText = rubric.steps
      .map((s, i) => `  ${i + 1}. ${s}`)
      .join("\n");
    sections.push(
      `<rubric name="${rubric.name}">\n` +
        `Criteria: ${rubric.criteria}\n` +
        `Evaluation steps:\n${stepsText}\n` +
        `</rubric>`,
    );
  }

  sections.push(
    "Now produce the JSON evaluation object for the rubric(s) above. " +
      "Use the exact rubric names as keys.",
  );

  return sections.join("\n\n");
}

File: plugins/pytest-agent-observability/src/pytest_agent_observability/judges/prompt.py
"""Build the combined LLM judge prompt for one or more rubrics."""

from __future__ import annotations

import json

from .rubrics import Rubric
from .types import JudgeInput


def build_prompt(rubrics: list[Rubric], input: JudgeInput) -> str:
    """Return a single prompt string instructing the model to evaluate all rubrics.

    The model must reply with JSON:
      {"<rubric_name>": {"score": float, "reason": str}, ...}
    """
    sections: list[str] = []

    sections.append(
        "You are an objective evaluator. Assess the agent response below against each "
        "rubric criterion and return ONLY a JSON object. No prose outside the JSON.\n"
        "Response format:\n"
        '{"<rubric_name>": {"score": <float 0.0-1.0>, "reason": "<brief explanation>"}, ...}'
    )

    # --- Input slots (omit empty) ---
    if input.system_prompt:
        sections.append(f"<system_prompt>\n{input.system_prompt}\n</system_prompt>")

    if input.task_instructions:
        sections.append(
            f"<task_instructions>\n{input.task_instructions}\n</task_instructions>"
        )

    if input.context:
        ctx = (
            "\n".join(input.context)
            if isinstance(input.context, list)
            else input.context
        )
        sections.append(f"<context>\n{ctx}\n</context>")

    if input.conversation_history:
        history_text = json.dumps(input.conversation_history, indent=2)
        sections.append(
            f"<conversation_history>\n{history_text}\n</conversation_history>"
        )

    sections.append(f"<agent_response>\n{input.response}\n</agent_response>")

    # --- Rubric definitions ---
    for rubric in rubrics:
        steps_text = "\n".join(f"  {i + 1}. {s}" for i, s in enumerate(rubric.steps))
        sections.append(
            f"<rubric name=\"{rubric.name}\">\n"
            f"Criteria: {rubric.criteria}\n"
            f"Evaluation steps:\n{steps_text}\n"
            f"</rubric>"
        )

    sections.append(
        "Now produce the JSON evaluation object for the rubric(s) above. "
        "Use the exact rubric names as keys."
    )

    return "\n\n".join(sections)
